Generic form components

andymikulski opened the pull request:
- ~Depends on PR #1121~

- Converts the `[Create|Edit|Clone][Recipe|Extension]Page` components over to the [template method pattern](https://en.wikipedia.org/wiki/Template_method_pattern)
  - All pages with forms inherit from `AbstractFormPage`, which handles common validation functionality which was otherwise duplicated across each component
   - Pages define a handful of 'abstract' functions (`getTitle`, `getFormComponent`, etc), which are called in `AbstractFormPage.render`

This cuts down on a lot of copy-pasted code across each page, however I wonder if the abstraction adds an unnecessary layer of complexity.

andymikulski commented:
Ready for review!

andymikulski commented on recipe-server/client/control/components/recipes/AbstractFormPage.js:
```diff
@@ -0,0 +1,77 @@
+  };
+
+  state = {
+    formErrors: undefined,
```
I played a little with overriding `setState` to prevent subclasses from changing `formErrors`, but that felt a little hacky. I've simply put a comment in `AbstractFormPage` mentioning that while the state _is_ availabe, `formErrors` should not touched.

mythmon replied:
I don't think I explained my concern correctly. Consider this subclass (nonsensical as it is):

```js
class StatefullFormPage extends AbstractFormPage {
  state = {
    subclassState: 42,
  }
  // ... required methods ...
}
```

Since this class sets its own state in the class definition, it would erase the state from AbstractFormPage, right? The same issue applies if the state is set in the constructor. I guess it doesn't matter in this case since the value is `undefined` either way.

andymikulski replied:
Ah! Good point. Interesting point. I will investigate.

andymikulski commented on recipe-server/client/control/components/recipes/AbstractFormPage.js:
```diff
@@ -0,0 +1,77 @@
+    formErrors: undefined,
+  };
+
+  /* hook */ onBeforeSubmit() {
```
Optionally abstract. The new JSDoc comments removes this note, though.

andymikulski commented on recipe-server/client/control/components/recipes/AbstractFormPage.js:
```diff
@@ -0,0 +1,77 @@
+    });
+  }
+
+  /* abstract */ onFailure() {
```
I've included the `@abstract` jsdoc prop to the comments, which defines the property as `[identifying] members that must be implemented (or overridden) by objects that inherit the member.` I think this should suffice.

andymikulski commented on recipe-server/client/control/components/recipes/AbstractFormPage.js:
```diff
@@ -0,0 +1,77 @@
+    try {
+      const actionValues = await this.performAction(formValues);
+
+      this.onSuccess(...actionValues);
```
I'd like to explore this a bit so I haven't touched it yet. The idea behind the spread operator was to keep functions written as `doThing(someVal, otherVal)` versus `doThing(vals){ someVal = vals[0]; otherVal = vals[1]; }`. I will test out some ideas and go from there.

mythmon replied:
Without spread, the second example could be written as `doThings([someVal, otherVal])`. It would also let users do things like `doThings({ someVal, otherVal })` (ie, passing an options object instead of ordered arguments), or a simple single value `doThings(someVal)`, neither of which would be allowed with the spread operator.

andymikulski commented:
Alright, I've done a bit more refactoring on this and simplified the approach, while making it a bit more Reacty.

Essentially, I've converted what was the `AbstractFormPage` into a `GenericFormContainer` component. This solves a handful of problems surrounding state, and alleviates some of the extra complexity which came out of the template approach.

This is ready for another pass, though requires some tests.

andymikulski commented:
Weird, Circle had some errors in the tests I wrote for the component. I didn't have this problem locally, so I'll have to inspect.

mythmon commented on recipe-server/client/control/components/recipes/GenericFormContainer.js:
```diff
@@ -0,0 +1,65 @@
+
+  /**
+   * Render does not need to be overridden in extended pages. If you choose to
+   * create a custom render function, be sure to call `super.render.call(this)`.
```
I think this comment doesn't belong here anymore?

mythmon commented on recipe-server/client/control/tests/components/recipes/test_GenericFormContainer.js:
```diff
@@ -0,0 +1,102 @@
+
+    wrapper.find(CustomEl).simulate('click');
+    // formAction is async so we need to wait for that promise to resolve.
+    setTimeout(() => {
```
setTimeout shouldn't be used in tests like this. The onSuccess callback should resolve a promise that this function can await on. Also, since this test is not async, the test will pass before this callback fires.

mythmon commented:
I like the new `GenericFormContainer` better. It's a clearer win, in terms of abstraction to me. Nice work with that. There are a couple problems, but I think this is close to ready.

mythmon commented on recipe-server/client/control/tests/components/recipes/test_GenericFormContainer.js:
```diff
@@ -65,24 +65,23 @@ describe('<GenericFormContainer>', () => {
     const CustomEl = ({ onSubmit }) => <div onClick={onSubmit}>Test</div>;
-    let success = false;
+
+    const success = new Promise(r => r(true));
```
This creates an immediately resolved promise, which means that the test below will always pass. A better way to do this would be 

```js
let resolve;
const success = new Promise(r => resolve = r);
onSuccess(() => { resolve(true); });
expect(await success).toBe(true);
```

I recommend making this test fail (maybe by changing the code, or maybe by removing the contents of the onSuccess handler) so you can be sure the solution you have is a valid test, instead of one that passes unconditionally.

mythmon commented:
What's the status of this? There still seem to be some test failures in CI.

andymikulski commented:
Circle was hitting errors which I was not getting locally. The issue stemmed from using `.find(CustomEl)` where CustomEl was a stateless functional component. I've switched over to targeting the element ID instead. Assuming Circle is happy with these changes, this PR should be good to go.

mythmon commented:
bors r+

mythmon commented:
Ah, taskcluster, never change. Lets ask bors to try again

bors r+
